distributed/active_memory_manager.py

```diff
@@ -0,0 +1,429 @@
+import asyncio
+import heapq
+import math
+from collections import defaultdict
+from collections.abc import Generator, Iterable, Iterator
+from typing import Optional
+
+import dask
+
+from .scheduler import SchedulerPlugin, SchedulerState, TaskState, WorkerState
+
+
+class ActiveMemoryManager(SchedulerPlugin):
+    scheduler: SchedulerState
+    plugins: "set[ActiveMemoryManagerPlugin]"
+    interval: float
+
+    # Daemon that runs every <interval> seconds
+    run_task: asyncio.Task
+
+    # These attributes only exist within the scope of self.run()
+    # Current memory (in bytes) allocated on each worker, plus/minus pending actions
+    workers_memory: "dict[WorkerState, int]"
+    # Pending replications and deletions for each task
+    pending: "defaultdict[TaskState, tuple[set[WorkerState], set[WorkerState]]]"
+
+    def __init__(
+        self,
+        plugins: "Iterable[ActiveMemoryManagerPlugin]" = None,
+        interval: float = 2.0,
+    ):
+        self.interval = interval
+        self.plugins = set()
+        if plugins is None:
+            plugins = {ReplicateReduce(), Rebalance()}
+        for plugin in plugins:
+            self.register_plugin(plugin)
+
+    def register_plugin(self, plugin: "ActiveMemoryManagerPlugin") -> None:
+        plugin.manager = self
+        self.plugins.add(plugin)
+
+    async def start(self, scheduler: SchedulerState):
+        """SchedulerPlugin hook. Start kicking off self.run every <interval> seconds
+        until scheduler shutdown.
+        """
+        self.scheduler = scheduler
+
+        async def _():
+            while True:
+                await self.run()
+                await asyncio.sleep(self.interval)
+
+        self.run_task = asyncio.create_task(_())
+
+    async def close(self):
+        """SchedulerPlugin hook. Stop kicking off self.run."""
+        self.run_task.cancel()
+
+    async def run(self):
+        """Run all plugins once and enact their recommendations to replicate/drop keys"""
+        self.workers_memory = {
+            w: w.memory.optimistic for w in self.scheduler.workers.values()
+        }
+        self.pending = defaultdict(lambda: (set(), set()))
+        self._run_plugins()  # populate self.pending
+
+        # TODO WIP act on self.pending
+
+        del self.workers_memory
+        del self.pending
+
+    def _run_plugins(self) -> None:
+        """Sequentially run ActiveMemoryManagerPlugin.run() of all registered plugins,
+        obtain replicate/drop suggestions, and use them to populate self.pending.
+        """
+        candidates: "Optional[set[WorkerState]]"
+        cmd: str
+        ws: WorkerState
+        ts: TaskState
+        nreplicas: int
+
+        for plugin in self.plugins:
+            plugin_gen = plugin.run()
+            while True:
+                try:
+                    cmd, ts, candidates = next(plugin_gen)
+                except StopIteration:
+                    break  # next plugin
```

Comment:
> However, it looks like that information is differently shaped. Maybe it should be a separate iterator?

I don't think I understand what you mean; could you articulate?

Reply:
Below I see that you call `plugin_gen.send(ts)`.  That would explain why you're calling next on this thing rather than using a for loop.  However, my brief look at the type of `ts` showed that it didn't fit the `cmd, ts, candidates` output expected here.  I only looked at this briefly though, and so could easily be wrong.

Reply:
it's ``plugin_gen.send(ws)``. the type of ws is ``Optional[WorkerState]``.

The signature of the object returned by the run() method is
```python
Generator[
    tuple[str, TaskState, Optional[set[WorkerState]]],
    Optional[WorkerState],
    None,
]
```